src/main/kotlin/org/rust/ide/intentions/CreateFunctionIntention.kt

```diff
@@ -15,123 +15,198 @@ import org.rust.ide.presentation.renderInsertionSafe
+import org.rust.lang.core.types.ty.*
 import org.rust.lang.core.types.type
 import org.rust.openapiext.buildAndRunTemplate
 import org.rust.openapiext.createSmartPointer
 
 class CreateFunctionIntention : RsElementBaseIntentionAction<CreateFunctionIntention.Context>() {
     override fun getFamilyName() = text
 
-    sealed class Context {
-        data class Function(val callExpr: RsCallExpr, val name: String, val module: RsMod) : Context()
+    sealed class Context(val name: String, val callElement: PsiElement) {
+        abstract val visibility: String
+        abstract val arguments: RsValueArgumentList
+        abstract val returnType: Ty?
+
+        class Function(val callExpr: RsCallExpr, name: String, val module: RsMod) : Context(name, callExpr) {
+            override val visibility: String = if (callExpr.containingMod != module) "pub(crate) " else ""
+            override val arguments: RsValueArgumentList = callExpr.valueArgumentList
+            override val returnType: Ty? = callExpr.expr.expectedType
+        }
+
+        class Method(val methodCall: RsMethodCall, name: String, val item: RsStructOrEnumItemElement)
+            : Context(name, methodCall) {
+            override val visibility: String
+                get() {
+                    val parentImpl = methodCall.parentOfType<RsImplItem>()
+                    // creating a method inside the same impl
+                    return if ((parentImpl?.typeReference?.type as? TyAdt)?.item == item && parentImpl.traitRef == null) {
+                        ""
+                    } else "pub(crate)"
+                }
+            override val arguments: RsValueArgumentList = methodCall.valueArgumentList
+            override val returnType: Ty? = methodCall.parentDotExpr.expectedType
+        }
     }
 
     override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
         val path = element.parentOfType<RsPath>()
-        val functionCall = path?.parentOfType<RsCallExpr>() ?: return null
-        if (!path.isUnresolved) return null
-
-        if (!functionCall.expr.isAncestorOf(path)) return null
-
-        val module = getTargetModuleForFunction(path) ?: return null
-        val name = path.referenceName
-        text = "Create function `$name`"
-        return Context.Function(functionCall, name, module)
+        val functionCall = path?.parentOfType<RsCallExpr>()
+        if (functionCall != null) {
+            if (!path.isUnresolved) return null
+            if (!functionCall.expr.isAncestorOf(path)) return null
+
+            val module = getTargetModuleForFunction(path) ?: return null
+            val name = path.referenceName
+            text = "Create function `$name`"
+            return Context.Function(functionCall, name, module)
+        }
+        val methodCall = element.parentOfType<RsMethodCall>()
+        if (methodCall != null) {
+            if (methodCall.reference.resolve() != null) return null
```

Comment:
Same as in function call case, let's limit possible caret position to the method identifier. `element == methodCall.identifier` should works